dev/benchmarks/macrobenchmarks/lib/src/web/recorder.dart

```diff
@@ -349,19 +384,22 @@ abstract class WidgetRecorder extends Recorder implements FrameRecorder {
   Profile profile;
   Completer<void> _runCompleter;
 
+  /// Whether to delimit warm-up frames in a custom way.
+  final bool useCustomWarmUp;
```

Comment:
What do you think of using a strategy pattern? We could have an abstract `WarmUpStrategy` class with default implementation, which could be replaced by a custom one. Then there would be no need to check `if (useCustomWarmUp) ...`. The API of the strategy class could be something like:

```dart
abstract class WarmUpStrategy {
  bool shouldContinueWarmUp(Profile profile);
}
```

For as long as `shouldContinueWarmUp` returns `true` the warm-up continues. As soon as it returns `false` we start measuring.

Reply:
There are some tricky issues we have to consider.

1. What API should we use to reconcile these two very different strategies?
    * In the **default** implementation, everything **except** the last 100 data points of each `Timeseries` is warm-up, so we are only able to determine the warm-up data points **after** we have all the data points recorded.
    * In the **new** implementation, the `Profile` marks data points as warm-up along the way.
2. Different `Timeseries` within a `Profile` may have different lengths.
3. Similarly, we are unable to skip warm-up **measured** frames (for "total UI frame") in the default implementation. Since we don't know how many data points we would have in the end, we cannot know where these "last 100 data points" begin, and therefore we cannot know when the warm-up ends.

I did a quick experiment. It turns out that all **existing** benchmarks record exactly 300 data points for each `Timeseries`, so the first 200 data points of each `Timeseries` are warm-up.
I suggest switching the **default** logic to using the **first 200** data points as warm-up. This should not change existing behavior, and it's easier to know which data points are warm-up as we go.

In this new setting, the default `shouldContinueWarmUp` would be
```dart
bool shouldContinueWarmUp (Profile profile) =>
    profile.scoreData.isEmpty || !profile.scoreData.values.every((String timeseries) => timeseries.count >= _kWarmUpSampleCount);
```
and the new `shouldContinueWarmUp` would be
```dart
bool shouldContinueWarmUp (Profile profile) => profile.isWarmingUp;
```

Reply:
Let's keep this as-is for now. We can introduce this in a later PR.